### ISSUE TITLE
manual DoF renumbering and TrilinosSolver issue with MPI

### ISSUE THREAD

**Issue by davydden** [dealii/dealii#2966]
with a reference to https://groups.google.com/forum/?fromgroups#!topic/dealii/ncEIt6y7EHg

Inverse of a mass matrix with Trilinos solvers gives:

- different results for 1 and 2 mpi cores with CG while matrix and RHS appears to be the same:
```
$ mpirun -np 2 test_renumbering
Start without renumbering
l1(mat) = 0.25
fr(mat) = 0.194444
l2(rhs) = 0.375
l2(rhs) = 0.375
l2(sol) = 3
Finished without renumbering
Start with renumbering
l1(mat) = 0.25
fr(mat) = 0.194444
l2(rhs) = 0.375
l2(rhs) = 0.375
l2(sol) = 5.96474
Finished with renumbering
```

- fails with 2 cores with direct solver

Attached is a slightly modified version of the file provided in the Google forums.
Could be a Trilinos bug, i was not able to spot any problems in the attached code.

[test_renumbering.cc.zip](https://github.com/dealii/dealii/files/417712/test_renumbering.cc.zip)

**Comment by drwells** [comment]
I get the same result with master and Trilinos 12.4.2. If I just use Cuthill McKee and comment out the rest of `renumber` then everything seems to work. I think that there is something fishy going on with this ad hoc renumbering algorithm (though the ranges written to deallog after it is run look reasonable).

**Comment by davydden** [comment]
@drwells i also checked `Cuthill McKee` and it works fine. In the Google thread it was suggested that Cuthill McKee works as it 

>  does the renumbering only on the locally owned dofs.

**Comment by drwells** [comment]
I really like this test case because it seems to have found at least three bugs.

I played around with this some more and it looks like the primary bug is in `shared::Triangulation<2>`. In particular, if one switches it out for a `distributed::Triangulation<2>` in the posted file then everything (including the direct solver) works correctly. I added a ton of logging and I think I figured out why (here is some of the output from processor zero):

```
[...]
DEAL::before renumbering:
DEAL::{[0,3]}
DEAL::after renumbering:
DEAL::{[5,8]}
DEAL::Done renumbering.
DEAL::locally owned dofs:
DEAL::{[5,8]}
DEAL::locally relevant dofs:
DEAL::{[0,8]}
[...]
DEAL::
DEAL::sparse matrix row range: 0, 4
```

so there is some disconnect between what the DoFHandler thinks it owns and what it tells the sparse matrix it owns (bug 1). When we write into this matrix the Trilinos sparse matrix wrappers don't complain at all but the PETSc ones correctly crash (bug 2).  Finally, I saved the actual values of the RHS and solution vectors:

```
DEAL::Convergence step 6 value 0
DEAL::rhs vector:
1.250e-01 6.250e-02 1.250e-01 6.250e-02 0.000e+00 2.297e-321 6.945e-310 6.945e-310 1.197e-316
DEAL::
DEAL::solution vector:
4.562e+00 -8.750e-01 5.687e+00 -1.250e+00 1.197e-316 2.297e-321 6.945e-310 6.945e-310 1.197e-316
```

so they are writing garbage (and the whole thing intermittently segfaults for me) (bug 3).

My repo is at

https://github.com/drwells/dealii-forum-trilinos-bug/

where I have a version without Trilinos (i.e., with PETSc).

**Comment by davydden** [comment]
thanks for digging it further.

>  there is some disconnect between what the DoFHandler thinks it owns and what it tells the sparse matrix it own

but how is it possible? After renumbering is done, the `IndexSet` objects you logged are used to initialize sparse matrix.

**Comment by bangerth** [comment]
It seems to me that the dofhandler objects are just a means to produce input objects that then subsequently confuse the matrix objects. Do you think you could come up with a more self contained test where you just fix these arguments to values you currently use? This may make the test smaller and easier to understand.

**Comment by drwells** [comment]
I added a minimal version to my test repository:

https://github.com/drwells/dealii-forum-trilinos-bug/tree/minimal

Regardless of the Triangulation that we use we get exceptions from Trilinos wrapper code if we refine the grid further, so it is just coincidence that `parallel::distributed::Triangulation` works with one level of refinement and `parallel::shared::Triangulation` does not.

**Comment by masterleinad** [comment]
Part of the problem is that we always construct a linear EpetraMap from an IndexSet. Therefore, we simply ignore the reordering. Working on a PR.

**Comment by drwells** [comment]
It is worth noting explicitly that this same bug also probably appears in the PETSc wrappers.

**Comment by masterleinad** [comment]
Yes, you are right. Since MatGetOwnershipRange (http://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Mat/MatGetOwnershipRange.html) just provides valid information for linear IndexSets, I don't see an easy way to fix this for PETSc. Hence, I would just assert that a linear IndexSet is used and tell that this feature is not implemented otherwise.

**Comment by masterleinad** [comment]
The output that I get after #3093 is:
> Start without renumbering
> [...]
> l1(mat) = 0.0625
> fr(mat) = 0.111111
> l2(rhs) = 0.21875
> l2(rhs) = 0.21875
> l2(sol) = 5
> Finished without renumbering
> Start with renumbering
> [...]
> l1(mat) = 0.0625
> fr(mat) = 0.110554
> l2(rhs) = 0.21875
> l2(rhs) = 0.21875
> l2(sol) = 5.15269
> Finished with renumbering

So there is still something wrong.

**Comment by drwells** [comment]
In the original example the ranges are still contiguous but the matrix ends up owning the 'wrong' set of rows, so there are still yet more uncovered bugs here.

**Comment by masterleinad** [comment]
Yes, but local_elements() now reports the correct elements (trilinos/renumbering_01).

**Comment by bangerth** [comment]
@masterleinad -- is this fixed with your PR #3093?

**Comment by masterleinad** [comment]
No, it's not. I still observe the same behavior as in https://github.com/dealii/dealii/issues/2966#issuecomment-246418577.
It seems that some elements for locally relevant DoFs get lost when writing to the global matrix.
